### PR TITLE
Abstract stylist to libcst style conversion

### PR DESCRIPTION
**Pull request by konstin**
Replace all duplicate invocations of

```rust
let mut state = CodegenState {
    default_newline: &stylist.line_ending(),
    default_indent: stylist.indentation(),
    ..CodegenState::default()
}
tree.codegen(&mut state);
state.to_string()
```

with

```rust
tree.codegen_stylist(&stylist)
```

No functional changes.

### CODE REVIEW

**Review by MichaReiser — APPROVED** [review]
Looks good to me. I don't know if it was intentional that the `ast` crate doesn't depend on the `cst` crate.

**Comment by konstin** [comment]
> I don't know if it was intentional that the ast crate doesn't depend on the cst crate.

i already though about making this a free standing function in the ruff crate otherwise, @charliermarsh any preferences?

**Comment by charliermarsh** [comment]
It wasn't intentional but it is a heavy dependency. Should we make it a feature?

**Comment by konstin** [comment]
To make this less hypothetical, this would be the alternative without the libcst dep: https://github.com/charliermarsh/ruff/pull/4777. If you want that version, you can merge it into this PR, if you don't want it you can close #4777 and merge this PR instead.

File: crates/ruff/src/autofix/codemods.rs [hunk 1]
```diff
@@ -11,119 +11,6 @@ use ruff_python_ast::source_code::{Locator, Stylist};
 use crate::cst::helpers::compose_module_path;
 use crate::cst::matchers::match_statement;
 
-/// Given an import statement, remove any imports that are specified in the `imports` iterator.
-///
-/// Returns `Ok(None)` if the statement is empty after removing the imports.
-pub(crate) fn remove_imports<'a>(
```
Comment (charliermarsh) [inline crates/ruff/src/autofix/codemods.rs:17]: Does this just need a rebase? Why was this deleted?

**Review by charliermarsh — APPROVED** [review]

File: crates/ruff/src/autofix/codemods.rs (hunk 1 above)
Comment (konstin) [reply crates/ruff/src/autofix/codemods.rs:17]: i think the rebase went wrong, i have to fix that

**Comment by charliermarsh** [comment]
Looks reasonable to me.

File: crates/ruff/src/autofix/codemods.rs [hunk 2]
```diff
@@ -11,6 +11,23 @@ use ruff_python_ast::source_code::{Locator, Stylist};
 use crate::cst::helpers::compose_module_path;
 use crate::cst::matchers::match_statement;
 
+/// Glue code to make libcst codegen work with ruff's Stylist
+pub(crate) trait CodegenStylist<'a>: Codegen<'a> {
+    fn codegen_stylist(&self, stylist: &'a Stylist<'a>) -> String;
```
Comment (MichaReiser) [inline crates/ruff/src/autofix/codemods.rs:16]: This seems to be sufficient (I honestly don't fully understand yet why lifetimes are required here)
```suggestion
    fn codegen_stylist(&self, stylist: &'a Stylist) -> String;
```

**Review by MichaReiser — APPROVED** [review]
